### PR TITLE
[vmem] Boundary tag allocator.

### PR DESCRIPTION
**Pull request by MichalBlk**
The primary problem with our current boundary tag allocation solution which is based on the pool allocator is that it creates a cyclic dependency between `pool` and `vmem`. If a pool happens to run out of free items in slabs, it calls `vmem` to allocate kernel virtual address space. However, if at that time `vmem` runs out of boundary tags, we reenter `pool_alloc` and panic due to an attempt to acquire a nonrecursive mutex for the second time. Moreover, changing the allocation scheme to allocate new slabs from some other area (e.g. `dmap`) doesn't fix all problems as there's still a cycle between vmem's and boundary tag pool's lock (which is sometimes detected by `lockdep`).

This PR contains changes introduced in #1327.

### CODE REVIEW

**Review by cahirwpz — COMMENTED** [review]
Please verify my changes to `kva_alloc` are sound and correct.

File: sys/kern/vmem.c [hunk 1]
```diff
@@ -75,16 +81,72 @@ typedef struct bt {
...
+      assert(page);
+
+      mtx_lock(&bt_lock);
+      bt_add_page(page);
+      bt_refiller = NULL;
+      cv_broadcast(&bt_cv);
+      break;
+    }
+
+    if (flags & M_NOWAIT)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/kern/vmem.c:132]: Done.

File: sys/kern/kmem.c [hunk 2]
```diff
@@ -12,22 +13,25 @@
...
-vaddr_t kva_alloc(size_t size) {
+vaddr_t kva_alloc(size_t size, kmem_flags_t flags) {
   assert(page_aligned_p(size));
-  vmem_addr_t start;
-  vaddr_t old = atomic_load(&vm_kernel_end);
+  vmem_addr_t start = 0;
+  int error;
 
   /*
    * Let's assume that vmem_alloc failed.
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/kern/kmem.c:37]: Yes, this is still up to date.

**Comment by MichalBlk** [comment]
@cahirwpz There was an issue with updating `max_kva`. You updated `max_kva` regardless of the result of `vmem_add`, thereby `if (max_kva == max_kva_prev)` was always true.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**